### ISSUE TITLE
Synology installation

### ISSUE THREAD

**Issue by frooonk** [NRCHKB/node-red-contrib-homekit-docker#6]
Hi,

when trying to install the NRCHKB-Docker-Image via the SSH-shell of my Synology DS220+ 

`sudo docker run it --net=host -v /volume1/docker/nrchkb:/data -e DSM_HOSTNAME=ds220 -e TZ=Europe/Berlin --name=homekit nrchkb/node-red-homekit`

I get the following error message:

`Unable to find image 'it:latest' locally
docker: Error response from daemon: pull access denied for it, repository does not exist or may require 'docker login'.
See 'docker run --help'.`

Do I have to add the described Environment variable before or after pulling the docker image?

Thanks in advance!

Frank

**Comment by oliverrahner** [comment]
Hi Frank,

you are missing a dash ('-') in your command line:

`sudo docker run **-**it --net=host -v /volume1/docker/nrchkb:/data -e DSM_HOSTNAME=ds220 -e TZ=Europe/Berlin --name=homekit nrchkb/node-red-homekit`

**Comment by frooonk** [comment]
Hi Oliver,

thank you very much. It works fine :-)

I copied the command (with the missing "-" before "it") from the website [https://hub.docker.com/r/nrchkb/node-red-homekit](url).

**Comment by frooonk** [comment]
Hi Oliver,

thanks for your fast response. 

Unfortunately I didn't notice the missing **-** in the code snippet when Copy&Paste...

A new error occured during the installation (see error message below). Can you give me a hint?

```
Unable to find image 'nrchkb/node-red-homekit:latest' locally
latest: Pulling from nrchkb/node-red-homekit
cbdbe7a5bc2a: Pull complete 
95feee427958: Pull complete 
4123295e9f39: Pull complete 
a59140832df1: Pull complete 
eb9ac9f35d0d: Pull complete 
52fa16aad946: Pull complete 
5db454e22d0e: Pull complete 
85adc26ae481: Pull complete 
d80ef11dbf56: Pull complete 
2594ef26041d: Pull complete 
81291faee776: Pull complete 
223ad3db3d6e: Pull complete 
f10a5fde7c0c: Pull complete 
029c814ea0fc: Pull complete 
855c9a41fa64: Pull complete 
c4176a476496: Pull complete 
ae263ccf54d1: Pull complete 
4300c4a0ba17: Pull complete 
7abc314e63d2: Pull complete 
ba5e5ccec9af: Pull complete 
Digest: sha256:378e52cf01f2af8c96521741db54a5cfada93104a7eb291bc5fa74aea8be4b60
Status: Downloaded newer image for nrchkb/node-red-homekit:latest
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] 55-node-red: applying... 
[fix-attrs.d] 55-node-red: exited 0.
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 40-dbus-avahi: executing... 
[cont-init.d] 40-dbus-avahi: exited 0.
[cont-init.d] 55-node-red: executing... 
[cont-init.d] 55-node-red: exited 0.
[cont-init.d] done.
[services.d] starting services
Starting node-red
Starting dbus-daemon
[services.d] done.

> node-red-contrib-homekit-docker@1.1.0 start /usr/src/node-red
> node $NODE_OPTIONS node_modules/node-red/red.js $FLOWS "--userDir" "/data"

Starting Avahi daemon
Found user 'avahi' (UID 86) and group 'avahi' (GID 86).
Successfully dropped root privileges.
avahi-daemon 0.7 starting up.
WARNING: No NSS support for mDNS detected, consider installing nss-mdns!
Loading service file /etc/avahi/services/sftp-ssh.service.
Loading service file /etc/avahi/services/ssh.service.
*** WARNING: Detected another IPv4 mDNS stack running on this host. This makes mDNS unreliable and is thus not recommended. ***
Joining mDNS multicast group on interface docker0.IPv4 with address 172.17.0.1.
New relevant interface docker0.IPv4 for mDNS.
Joining mDNS multicast group on interface bond0.IPv4 with address 192.168.50.11.
New relevant interface bond0.IPv4 for mDNS.
Network interface enumeration completed.
Registering new address record for fe80::42:b4ff:fe33:56bd on docker0.*.
Registering new address record for 172.17.0.1 on docker0.IPv4.
Registering new address record for fe80::211:32ff:fecc:69d9 on bond0.*.
Registering new address record for 192.168.50.11 on bond0.IPv4.
fs.js:114
    throw err;
    ^

Error: EACCES: permission denied, copyfile '/usr/src/node-red/node_modules/node-red/settings.js' -> '/data/settings.js'
    at Object.copyFileSync (fs.js:1728:3)
    at copyFile (/usr/src/node-red/node_modules/fs-extra/lib/copy-sync/copy-sync.js:68:8)
    at onFile (/usr/src/node-red/node_modules/fs-extra/lib/copy-sync/copy-sync.js:53:25)
    at getStats (/usr/src/node-red/node_modules/fs-extra/lib/copy-sync/copy-sync.js:48:44)
    at startCopy (/usr/src/node-red/node_modules/fs-extra/lib/copy-sync/copy-sync.js:38:10)
    at handleFilterAndCopy (/usr/src/node-red/node_modules/fs-extra/lib/copy-sync/copy-sync.js:33:10)
    at Object.copySync (/usr/src/node-red/node_modules/fs-extra/lib/copy-sync/copy-sync.js:26:10)
    at Object.<anonymous> (/usr/src/node-red/node_modules/node-red/red.js:125:20)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-red-contrib-homekit-docker@1.1.0 start: `node $NODE_OPTIONS node_modules/node-red/red.js $FLOWS "--userDir" "/data"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-red-contrib-homekit-docker@1.1.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /usr/src/node-red/.npm/_logs/2020-10-04T18_30_22_816Z-debug.log
Starting node-red
Server startup complete. Host name is ds220.local. Local service cookie is 86090614.

> node-red-contrib-homekit-docker@1.1.0 start /usr/src/node-red
> node $NODE_OPTIONS node_modules/node-red/red.js $FLOWS "--userDir" "/data"

fs.js:114
    throw err;
    ^
````

**Comment by Shaquu** [comment]
@frooonk sorry for not so fast response.
It looks like your user running docker doesn't have permissions to write to file data/settings,js is

**Comment by frooonk** [comment]
Hi Shaquu, thanks for response. I had a lot to do anyway...

You showed me the right direction: it must have something to do with the data-folder. If I change the command-line like this, the container and node-red is starting on my Synology-NAS:

`sudo docker run -it --net=host -v /volume1/docker/nrchkb:/volume1/docker/data -e DSM_HOSTNAME=ds220 -e TZ=Europe/Berlin --name=homekit nrchkb/node-red-homekit`

But in Node-Red I get a message to map a persistent data-volume. So it doesn't work properly. It seems like I have no idea how to correctly create a persistent data volume in my NAS-directory /volume1/docker using the command line...

The next thing is, that I can't see or add a new device (Bridge) in the HomeKit app although I specified the option --net=host in the command-line.

**Comment by oliverrahner** [comment]
Hi Frank, you need to keep the part after the colon ":" unmodified for the -v parameter, like you had before.
The issue is that User Id 1000, as used by the docker image, needs access to the directory `/volume1/docker/nrchkb`. This can be done as mentioned in the [Readme](https://github.com/NRCHKB/node-red-contrib-homekit-docker#permissions):

````
chown -R  1000:1000 /volume1/docker/nrchkb
````

**Comment by frooonk** [comment]
Thanks Oliver! Now it works :-)

Just for my understanding: after deploying a new flow in NodeRed, the flows now will be stored persistant in the directory /volume1/docker/nrchkb/data ???

I added a simple Example Bridge with one switch just for testing and the only folder that was updated is /volume1/docker/nrchkb/homekit-persist !!!

I have a lot of NodeRed-Flows and I want to be sure, that all modified flows will still be there after a NRCHKB-update...

**Comment by oliverrahner** [comment]
You only got it slighly wrong.
You told docker to map `/data` to `/volume1/docker/nrchkb`, so there's no need for `data` to appear as a separate sub directory.
Your flows should appear in a json file named somthing like `flows-<hostname>.json` directly under `/volume1/docker/nrchkb/`.

**Comment by frooonk** [comment]
Hi Oliver, if I got you right, then `-v /volume1/docker/nrchkb:/data` will mount the directory "/data" inside the container to the directory "/volume1/docker/nrchkb" on the NAS. 

The content of the the mounted directory is the persistent data (homekit-persist, lib, node_modules and all the files like flows.json) which should be backup'ed on the NAS.

By the way a hint for all docker-beginners like me, I found an interesting Tutorial which shows how to create docker containers especially on Synology NAS. The only thing to do manually is to set the permissions (via command-line : `ls -nal <path_on_host>` and `chown -R  1000:1000 <path_on_host>`) and to set the environment variable named DSM_HOSTNAME (via the docker-settings of the container):

[https://www.baitando.com/it/2019/09/22/using-docker-on-synology-nas](url)

Thanks again!
